### ISSUE TITLE
Extending `response_fields` syntax to allow rejecting/adding individual fields relative to the default

### ISSUE THREAD

**Issue by ml-evs** [Materials-Consortia/OPTIMADE#402]
I think a potentially useful piece of syntax that we are missing is the ability to specify `response_fields` relative to the default response for that implementation.

Typically you run into this when you require e.g., one extra field but would still like to access all the provider data, for example, for the OQMD, you need to manually request `species`, but don't want to have to list the other 50+ fields served by the API. You can also imagine a case where you want everything *but* `cartesian_site_positions`, which maybe prohibitively large, which would also require listing many many fields. 

I propose we add the following syntax (following `sort`) to handle these two cases:

- `?response_fields=+species`
    - Return all default fields, but also include `species` if not already present.
- `?response_fields=-cartesian_site_positions`
    - Return all default fields, but exclude `cartesian_site_positions` if present.
- `?response_fields=+species,-cartesian-site_positions`
    - Hopefully obvious.
- `?response_fields=+species,lattice_vectors`
    - Any mixed syntax like the above should probably just strip any fields with the `-` or `+` syntax and return those without the prefix.

Is this something we think would be generally useful and readily implementable?

A natural extension could then be to request whole *namespaces* as includes/excludes, like:

`?response_fields=+_biomol` (since `_biomol` on its own is not a valid field name) or perhaps with some wildcard syntax: `?response_fields=-_biomol_*` to make it more obvious.

**Comment by JPBergsma** [comment]
This could be useful, especially when people query the database manually and they do not want to have to type the entire list of properties that need to be returned. 
Should we also add a list of which properties are shared by default to info endpoint? Now you can only determine which fields are shared by default, by first sending a dummy query, and seeing which fields are returned.

**Comment by ml-evs** [comment]
> ~Should we also add a list of which properties are returned by default to info endpoint? Now you can only determine which fields are shared by default, by first sending a dummy query, and seeing which fields are returned.~ (If I read section 5.3 of the OPTIMADE standard correctly, it is already determined by the standard which properties are returned by default. Although not all database providers seem to adhere to this.)

I think only `id`, `type` and `last_modified` are strictly required (REQUIRED) in the default response, which most implementations seem to do fine with. Everything else is optional (which I think is fine).

I guess the only concern here is that the default fields returned by an implementation can change whenever they want, so queries using this syntax might not be perfectly repeatable.

**Comment by JPBergsma** [comment]
I had interpreted REQUIRED as also containing all the MUST and SHOULD properties. But it seems you are right and only the properties explicitly marked as REQUIRED should be returned.

In the OPTIMADE standard it says: "Properties not marked as REQUIRED in any case, MUST be included only if explicitly requested in the query parameter `response_fields`. Otherwise they SHOULD NOT be included." 

Many of the Optimade properties are not "REQUIRED". They are instead labelled as MUST or SHOULD be supported. 
So basically every OPTIMADE server does not adhere to the standard when it automatically returns fields like `cartesian_site_positions` or `elements`. 
I am not saying that this is a bad behaviour, but we should adjust the standard.

**Comment by ml-evs** [comment]
> In the OPTIMADE standard it says: "Properties not marked as REQUIRED in any case, MUST be included only if explicitly requested in the query parameter `response_fields`. Otherwise they SHOULD NOT be included."
> Many of the Optimade properties are not "REQUIRED". They are instead labelled as MUST or SHOULD be supported. So basically every OPTIMADE server does not adhere to the standard when it automatically returns fields like `cartesian_site_positions` or `elements`. I am not saying that this is a bad behaviour, but we should adjust the standard.

I think in context, this part is trying to say that when someone asks for `response_fields=cartesian_site_positions`, you should not include, say,  `elements` (which leads to the issue above). It is not making any point about what fields are returned when `response_fields` is not present.

Perhaps this could be clearer in the spec though.

**Comment by JPBergsma** [comment]
As you mentioned in" Add ASE-based visualisation usnistgov/nmrr-optimade-demo#4" not all the databases now offer all the MUST and SHOULD fields of the structures.    
So I think we should indeed write it more clearly in the spec.

**Comment by ml-evs** [comment]
I think I'm still keen on this functionality (both syntactically and logically) (though have just stumbled across #181 where @merkys and @rartino already hashed out pretty much the same idea)!
